Code for the first task. Class implementation-priority queue

D-mercool opened the pull request:
Попробовал реализовать класс, с заданной структурой данных, в моем случаем - очередь с приоритетом. Сделал меню, для демонстрации методов

D-mercool committed:
Code for the first task. Class implementation-priority queue

D-mercool committed:
Add function push_elements() and fixed a bug with an empty queue

Const-Phi commented on .ipynb_checkpoints/task1-checkpoint.ipynb:
```diff
@@ -0,0 +1,299 @@
+    "            print('Вы ввели не верный номер. Попробуйте еще раз')\n",
+    "            continue\n",
+    "\n",
+    "        if number == 1:\n",
+    "            queue = create_queue()\n",
+    "        elif number == 2:\n",
+    "            if queue == 'None':\n",
+    "                print('\\nОчерь не создана! Создайте очередь с приоритетом.')\n",
+    "            else:\n",
+    "                push_element(queue)  \n",
+    "        elif number == 3:\n",
+    "            if queue == 'None':\n",
+    "                print('\\nОчерь не создана! Создайте очередь с приоритетом.')\n",
+    "            else:\n",
+    "                pop_element(queue) \n",
+    "        elif number == 4:\n",
+    "            if queue == 'None':\n",
+    "                print('\\nОчерь не создана! Создайте очередь с приоритетом.')\n",
+    "            else:\n",
+    "                queue.print()\n",
+    "        elif number == 5:\n",
+    "            if queue == 'None':\n",
+    "                print('\\nОчерь не создана! Создайте очередь с приоритетом.')\n",
+    "            else:\n",
+    "                replace_element(queue)\n",
+    "        elif number == 6:\n",
+    "            if queue == 'None':\n",
+    "                print('\\nОчерь не создана! Создайте очередь с приоритетом.')\n",
+    "            else:\n",
+    "                push_elements(queue)\n",
```
Я бы несколько переписал этот код, повысив повторное использование посредством повышения уровня абстракции &ndash; использование [&lambda;-функций](https://realpython.com/python-lambda/) и [функции высшего порядка](https://ru.wikipedia.org/wiki/Функция_высшего_порядка) нам в помощь. Как образец приведу следующий отрывок кода
```python
def wrapper(queue, target):
  if queue == 'None':
    print('\\nОчерь не создана! Создайте очередь с приоритетом.')
  else:
    target(queue)

def print_queue(queue):
  queue.print()

def user_interface():
    # ...

    if number == 1:
        queue = create_queue()
    elif number == 2:
        wrapper(queue, push_element)  
    elif number == 3:
        wrapper(queue, pop_element) 
    elif number == 4:
        wrapper(queue, print_queue)
    elif number == 5:
        wrapper(queue, replace_element)
    elif number == 6:
        wrapper(queue, push_elements)

    # ...
```

D-mercool replied:
Пока что в программу можно вводить только инты, если реализовать метод для ввода стрингов, то тоже будет работать. В случаем со стрингами, приоритет по получит строка с Первой буквой по алфавиту. Объекты вводить не пробовал, и нужно ли это вообще?
По поводу .py, скидывать в этом формате.
Все комментарии понял, попробую сделать )

D-mercool replied:
Вопрос такой: Марина Алексеевна сказала не делать никаких выводов в консоль в классе. Придерживаться этого? стоит менять:
`class PriorityQueue():
    def print(self):
         print(self.heap)
`

D-mercool replied:
В данной структуре данных могут быть либо инты, либо стринги, вместе быть не могут, так как их не получится сравнивать между собой. НО можно реализовать структуру таким образом, что она будет хранить числа, как строку и строки, как строки. Например будет выглядеть так ['1', '2', '3', 'Alex', 'Dima']. Наивысший приоритет у наименьшего числа, затем идут строки, где наивысший приоритет определяется по алфавиту.

Const-Phi replied:
> Вопрос такой: Марина Алексеевна сказала не делать никаких выводов в консоль в классе. Придерживаться этого? стоит менять:
     `class PriorityQueue(): def print(self): print(self.heap)`

Короткий ответ следующий. Так как делаем &laquo;по аджайлу&raquo;, то имеет смысл обеспечить хоть какой-то способ вывода информации о внутреннем состоянии коллекции. Если это можно сделать малой кровью и хорошо, то делаем хорошо. Теоретически я бы добавил
```py
    def __str__(self):
    # что-то делаем для строкового представления коллекции
```

> В данной структуре данных могут быть либо инты, либо стринги, вместе быть не могут, так как их не получится сравнивать между собой. [...] 

И слава Богу! Так и должно быть!

> НО можно реализовать структуру таким образом, что она будет хранить числа, как строку и строки, как строки. Например будет выглядеть так ['1', '2', '3', 'Alex', 'Dima']. Наивысший приоритет у наименьшего числа, затем идут строки, где наивысший приоритет определяется по алфавиту.

Нет, подразумевалось, что хранится будут данные одного типа, но структуре всё равно какого &ndash; числа, строки, объекты, други коллекции etc.

Const-Phi commented on .ipynb_checkpoints/task1-checkpoint.ipynb:
```diff
@@ -0,0 +1,299 @@
+    "    queue = 'None'\n",
+    "    while True:\n",
+    "        try:\n",
+    "            number = int(input('Введите номер пункта меню: '))\n",
```
Кажется, что здесь можно тоже (_творчески переосмыслив_) использовать функцию `enter_element(text_input)`.

Const-Phi commented on .ipynb_checkpoints/task1-checkpoint.ipynb:
```diff
@@ -0,0 +1,299 @@
+    "\n",
+    "def create_queue():\n",
+    "    size = enter_element('\\nВведите размер очереди (целое положительно число): ')\n",
+    "    values = get_random_values(size)\n",
```
Кажется, что стоит разнести этот пункт меню на два &ndash; &laquo;Создать _пустую_ очередь&raquo; и &laquo;Создать очередь, _заполненную псевдослучайными значениями_&raquo;.

**UPD**
Как вариант второй (под)пункт можно свести к &laquo;Заполнить очередь  псевдослучайными значениями&raquo;, с созданием соответствующей функции и использования `wrapper`-а по этому знаменательному поводу.

Const-Phi reviewed (changes requested):
Не касаясь пока самой реализации структуры, помимо уже указанных замечаний, есть ещё два, носящих более-менее глобальный характер:

1. Подразумевается, что данная коллекция (как реализация структуры данных) позволяет хранить объекты _любого_ типа. _Так ли это?_
1. Сквозь формат `Jupiter`-а очень сложно продираться, не работая в нём (особенно при помощи текстового представления `GitHub`-а). _Существует ли альтернатива публиковать только код вида `*.py`, а не `*.ipynb`?_ На сколько мне видится, единственный выигрыш в последнем &ndash; наличие _как бы_ `test case`-ов.

kekafon commented:
Нужно ли добавлять метод для добавления строки в очередь?

D-mercool committed:
Add method create_empty_queue().
Add method wrapper(queue, target).
+ refactoring code

D-mercool committed:
Add task1.ipynb. task1.ipynb = task1.py
